platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/PropertyProblem.kt

```diff
@@ -142,149 +146,120 @@ sealed class PropertyTrace {
         override fun toString(): String = asString()
+        override fun describe(builder: StructuredMessage.Builder) = builder.apply {
+            reference(type.name)
+            text(" bean found in ")
+        }
     }
 
     data class Property(
         val kind: PropertyKind,
         val name: String,
         val trace: PropertyTrace
     ) : PropertyTrace() {
-        override val containingUserCode: String
-            get() = trace.containingUserCode
-
+        override val containingUserCodeMessage: StructuredMessage
+            get() = trace.containingUserCodeMessage
         override fun toString(): String = asString()
+        override fun describe(builder: StructuredMessage.Builder) = builder.apply {
+            text("$kind ")
+            reference(name)
+            text(" of ")
+        }
     }
 
     data class Project(
         val path: String,
         val trace: PropertyTrace
     ) : PropertyTrace() {
-        override val containingUserCode: String
-            get() = trace.containingUserCode
-
+        override val containingUserCodeMessage: StructuredMessage
+            get() = trace.containingUserCodeMessage
         override fun toString(): String = asString()
+        override fun describe(builder: StructuredMessage.Builder) = builder.apply {
+            text("project ")
+            reference(path)
+            text(" in ")
+        }
     }
 
     data class SystemProperty(
         val name: String,
         val trace: PropertyTrace
     ) : PropertyTrace() {
-        override val containingUserCode: String
-            get() = trace.containingUserCode
-
+        override val containingUserCodeMessage: StructuredMessage
+            get() = trace.containingUserCodeMessage
         override fun toString(): String = asString()
+        override fun describe(builder: StructuredMessage.Builder) = builder.apply {
+            text("system property ")
+            reference(name)
+            text(" set at ")
+        }
     }
 
-    abstract override fun toString(): String
-
     abstract override fun equals(other: Any?): Boolean
 
     abstract override fun hashCode(): Int
 
+    abstract override fun toString(): String
+
     /**
      * The shared logic for implementing `toString()` in subclasses.
      */
     protected
     fun asString(): String =
-        StringBuilder().apply {
+        StructuredMessage.Builder().apply {
             sequence.forEach {
-                appendStringOf(it)
+                it.describe(this)
             }
-        }.toString()
+        }.build().asString('`')
 
     /**
      * The user code where the problem occurred. User code should generally be some coarse-grained entity such as a plugin or script.
      */
     open val containingUserCode: String
-        get() = StringBuilder().apply {
-            appendStringOf(this@PropertyTrace)
-        }.toString()
-
-    private
-    fun StringBuilder.appendStringOf(trace: PropertyTrace) {
```

Comment:
Replaced by the various `describe(...)` overrides in the `PropertyTrace` subclasses